File: benchmark/BenchLarge.java
import generated_bench.Vec3Gen.*;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.io.File;

public class BenchLarge {
    public static void main(String[] args) throws Exception {
        String fileName = "large_payload.bin";
        File file = new File(fileName);
        if (!file.exists()) {
            // Try to find it in root if running from elsewhere
            if (new File("../" + fileName).exists()) {
                fileName = "../" + fileName;
            } else {
                 System.err.println("File not found: " + fileName);
                 System.exit(1);
            }
        }
        
        byte[] data = Files.readAllBytes(Paths.get(fileName));
        System.out.println("Read " + data.length + " bytes from " + fileName);

        // Warmup (optional, but good for Java)
        // For large data, maybe 1 run is enough for "cold" performance, but JIT matters.
        // Let's just run once for now.

        // Decode
        long start = System.nanoTime();
        WorldState world = WorldState.decode(data);
        long end = System.nanoTime();
        double elapsedDecode = (end - start) / 1e9;
        System.out.println(String.format("Decode time: %.4fs", elapsedDecode));

        System.out.println("Decoded " + world.guilds.length + " guilds");

        // Encode
        start = System.nanoTime();
        byte[] encodedData = world.encode();
        end = System.nanoTime();
        double elapsedEncode = (end - start) / 1e9;
        System.out.println(String.format("Encode time: %.4fs", elapsedEncode));
        System.out.println("Encoded size: " + encodedData.length + " bytes");

        if (encodedData.length != data.length) {
            System.out.println("WARNING: Encoded size mismatch! Original: " + data.length + ", New: " + encodedData.length);
        }
    }
}

File: benchmark/bench_large.js
const fs = require('fs');
const path = require('path');
const { WorldState } = require('./generated/js_bench/js/vec3');

const fileName = 'large_payload.bin';
let filePath = path.join(__dirname, '..', fileName);

if (!fs.existsSync(filePath)) {
    console.error(`File not found: ${filePath}`);
    process.exit(1);
}

const data = fs.readFileSync(filePath);
console.log(`Read ${data.length} bytes from ${fileName}`);

// Decode
const startDecode = process.hrtime();
const world = WorldState.decode(data);
const endDecode = process.hrtime(startDecode);
const elapsedDecode = endDecode[0] + endDecode[1] / 1e9;
console.log(`Decode time: ${elapsedDecode.toFixed(4)}s`);

console.log(`Decoded ${world.guilds.length} guilds`);

// Encode
const startEncode = process.hrtime();
const encodedData = world.encode();
const endEncode = process.hrtime(startEncode);
const elapsedEncode = endEncode[0] + endEncode[1] / 1e9;
console.log(`Encode time: ${elapsedEncode.toFixed(4)}s`);
console.log(`Encoded size: ${encodedData.length} bytes`);

if (encodedData.length !== data.length) {
    console.log(`WARNING: Encoded size mismatch! Original: ${data.length}, New: ${encodedData.length}`);
}
